File: Leetcode/0049-GroupAnagrams/cpp/solution1.cpp
#include <string>
#include <unordered_map>
#include <vector>

using namespace std;

class Solution {
 public:
  vector<vector<string>> groupAnagrams(vector<string>& strs) {
    unordered_map<string, vector<string>> map;
    for (const string& str : strs) {
      string key(26, 0);
      for (char c : str) {
        key[c - 'a']++;
      }
      map[key].push_back(str);
    }

    vector<vector<string>> res;
    for (auto& [key, val] : map) {
      res.push_back(val);
    }
    return res;
  }
};

File: Leetcode/0049-GroupAnagrams/java/Solution1.java
public class Solution1 {
    public List<List<String>> groupAnagrams(String[] strs) {
        Map<String, List<String>> anagrams = new HashMap<>();
        for (String s : strs) {
            // Generate a key for each string
            byte[] count = new byte[26];
            for (char c : s.toCharArray()) {
                count[c-'a']++;
            }
            String key = new String(count);
            if (!anagrams.containsKey(key)) {
                anagrams.put(key, new ArrayList<String>());
            }
            anagrams.get(key).add(s);
        }
        List<List<String>> result = new ArrayList<>();
        for (Map.Entry<String, List<String>> entry : anagrams.entrySet()) {
            result.add(entry.getValue());
        }
        return result;
    }
}